troi/playlist.py

```diff
@@ -1,79 +1,148 @@
-                'track_metadata' : {
-                    'artist_name' : e.artist.name if e.artist else "",
-                    'track_name' : e.name,
-                    'release_name' : e.release.name if e.release else "",
-                    'additional_info' : {
-                        'recording_mbid' : str(e.mbid),
-                        'artist_mbids' : artist_mbids
-                    }
-                }
-            })
-
-        self.entities = entities
-        self.playlist = ujson.dumps(listens, indent=4, sort_keys=True)
+        outputs = []
+        for input in inputs:
+            if isinstance(input[0], Recording):
+                if not is_homogeneous(input):
+                    raise TypeError("entity list not homogeneous")
+                self.playlists.append(Playlist(recordings=input))
+            elif isinstance(input[0], Playlist):
+                if not is_homogeneous(input):
+                    raise TypeError("entity list not homogeneous")
+                self.playlists.extend(input)
+            else:
+                raise PipelineError("Playlist passed incorrect input types.")
 
-        return entities
+        return outputs
 
     def print(self):
 
-        if not self.playlist:
-            print("[no playlist generated yet]")
+        if not self.playlists:
+            print("[no playlist(s) generated yet]")
             return
 
-        for recording in self.entities:
-            if not recording:
-                print("[invalid Recording]")
-                continue
-
-            if recording.artist is None or recording.artist.name is None:
-                artist = ""
-            else:
-                artist = recording.artist.name
-            if recording.name is None:
-                rec_name = ""
+        for i, playlist in enumerate(self.playlists):
+            if playlist.name:
+                print("playlist: '%s'" % playlist.name)
             else:
-                rec_name = recording.name
-            print("%-60s %-50s" % (rec_name[:59], artist[:49]))
+                print("playlist: %d" % i)
+
+            for recording in playlist.recordings:
+                if not recording:
+                    print("[invalid Recording]")
+                    continue
+
+                if recording.artist is None or recording.artist.name is None:
+                    artist = ""
+                else:
+                    artist = recording.artist.name
+                if recording.name is None:
+                    rec_name = ""
+                else:
+                    rec_name = recording.name
+                print("%-60s %-50s %s" % (rec_name[:59], artist[:49], str(recording.year or "")))
+            print()
+
+
+    def save(self):
+
+        if not self.playlists:
+            raise PipelineError("Playlists have not been generated yet.")
+
+        for i, playlist in enumerate(self.playlists):
+            filename = playlist.filename or "playlist %03d.jspf" % i
```

Comment:
can we have no spaces in this filename, to make it easy to select from a terminal?

Reply:
done